kernel/work.c

```diff
@@ -0,0 +1,971 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Second generation work queue implementation
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <spinlock.h>
+#include <errno.h>
+#include <ksched.h>
+#include <sys/printk.h>
+
+/* Lock to protect the internal state of all work items, and
+ * pending_cancels.
+ *
+ * No queue locks may be held when this is taken, but a queue lock may
+ * be taken while this is held.
+ */
+static struct k_spinlock lock;
+
+/* Invoked by work thread */
+static void handle_flush(struct k_work *work)
+{
+	struct z_work_flusher *flusher
+		= CONTAINER_OF(work, struct z_work_flusher, work);
+
+	k_sem_give(&flusher->sem);
+}
+
+static inline void init_flusher(struct z_work_flusher *flusher)
+{
+	k_sem_init(&flusher->sem, 0, 1);
+	k_work_init(&flusher->work, handle_flush);
+}
+
+/* List of pending cancellations. */
+static sys_slist_t pending_cancels;
+
+/* Initialize a canceler record and add it to the list of pending
+ * cancels.
+ *
+ * Invoked with work lock held.
+ *
+ * @param canceler the structure used to notify a waiting process.
+ * @param work the work structure that is to be canceled
+ */
+static inline void init_work_cancel(struct z_work_canceller *canceler,
+				    struct k_work *work)
+{
+	k_sem_init(&canceler->sem, 0, 1);
+	canceler->work = work;
+	sys_slist_append(&pending_cancels, &canceler->node);
+}
+
+/* Complete cancellation of a work item and unlock held lock.
+ *
+ * Invoked with work lock held.
+ *
+ * Invoked from a work queue thread.
+ *
+ * Reschedules.
+ *
+ * @param work the work structre that has completed cancellation
+ */
+static void finalize_cancel_locked(struct k_work *work)
+{
+	struct z_work_canceller *wc, *tmp;
+	sys_slist_t cancels;
+
+	/* Clear this first, so released high-priority threads don't
+	 * see it when doing things.
+	 */
+	atomic_clear_bit(&work->flags, K_WORK_CANCELING_BIT);
+
+	/* Move all pending cancels associated with this work item to
+	 * another list, under lock.  (We can't signal the release
+	 * while the lock is held because it's a reschedule point,
+	 * which would release the lock.)
+	 */
+	sys_slist_init(&cancels);
```

Comment:
You are not doing it, are you ? AFAIU you are doing the opposite, removing from the pending list and signaling from here, locked.

Reply:
You're right; this was based on an earlier incorrect belief that the next thread to run might be changed by `k_sem_give()`.  #29800 corrects the documentation that led to this belief, and I changed the implementation to match, but didn't update the description.